python/cog/predictor.py

```diff
@@ -163,19 +163,38 @@ def load_slim_predictor_from_file(
     return module
 
 
-def get_predictor(module: types.ModuleType, class_name: str) -> Any:
+def get_predictor(
+    module: types.ModuleType,
+    class_name: str,
+    *,
+    emit_metric: Optional[Callable[[str, Union[float, int]], None]] = None,
+) -> Any:
     predictor = getattr(module, class_name)
     # It could be a class or a function
     if inspect.isclass(predictor):
-        return predictor()
+        if emit_metric is None:
+
+            def noop_emit_metric(name: str, value: Union[float, int]) -> None:  # pylint: disable=unused-argument
+                pass
+
+            emit_metric = noop_emit_metric
+
+        try:
```

Comment:
This shenanigans feels a little risky, given that it's interfering directly with the user's code via `BasePredictor`.

I know generally thread-local state bad, etc., but in this case I think I actually marginally prefer the approach taken by the `async` branch in which `emit_metric` is just a function you import from somewhere else in Cog, and the details of how predictions are matched up to the call are up to us.

Reply:
Maybe it's helpful to be a bit concrete about how that would work. I would imagine that `emit_metric` is a module function which uses a `worker` ContextVar which holds a [`weakref`](https://docs.python.org/3/library/weakref.html) (probably a `weakref.proxy`) to the current child worker. 

Then `emit_metric` would call `worker_proxy.emit_metric(m)` and handle `ReferenceError` -- probably logging a warning if it happens.

I _think_ that this would generalise to the future with concurrent/async predictions, because `worker.emit_metric` could use its own `ContextVar` to track which prediction called it.